Added support for readonly properties

michaelpetri opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Requested with https://github.com/laminas/laminas-code/issues/101
Fixes #101

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -159,6 +171,47 @@ public function isConst()
 
+    /**
+     * @param  bool $readonly
+     * @return PropertyGenerator
```
Since we are on PHP 7.4+, we can declare these types natively (with `self` and `bool`) - yes, aware that the other methods don't have that :)

michaelpetri replied:
Added types all over in all files i touched.

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -159,6 +171,47 @@ public function isConst()
+    {
+        if ($readonly) {
+            $this->setFlags(self::FLAG_READONLY);
+        } else {
```
I suggest using an early return in general

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -159,6 +171,47 @@ public function isConst()
+    /**
+     * {@inheritDoc}
+     */
+    public function setFlags($flags)
```
Flags are quite a low-level representation of this: let's make this method `private`, if possible

michaelpetri replied:
This is just an override of the parent in order to throw exceptions for impossible flag combinations

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -159,6 +171,47 @@ public function isConst()
+            throw new RuntimeException('Modifier "readonly" in combination with "constant" not permitted.');
+        }
+
+        return parent::setFlags($flags);
```
Hmm, wait, the parent class also had flags? :O

Ocramius commented on test/Generator/PropertyGeneratorTest.php:
```diff
@@ -330,4 +380,20 @@ public function testFromReflectionOmitsTypeHintInTypedProperty(): void
+
+    public function testFromReflectionReadonlyProperty(): void
+    {
+        if (false === strpos((string)phpversion(), '8.1')) {
```
This can be `/** @requires php>=8.1 */` at docblock level :+1:

michaelpetri replied:
Syntax is `/** @requires PHP >= 8.1 */`

Ocramius commented on test/Generator/PropertyGeneratorTest.php:
```diff
@@ -330,4 +380,20 @@ public function testFromReflectionOmitsTypeHintInTypedProperty(): void
+        $generator = PropertyGenerator::fromReflection($reflectionProperty);
+        $code      = $generator->generate();
+
+        self::assertSame('    private $typedProperty;', $code);
```
This assertion seems to indicate that the `readonly` flag was not added? Seems wrong

michaelpetri commented on src/Generator/PropertyGenerator.php:
```diff
@@ -112,18 +121,21 @@ public static function fromArray(array $array)
-     * @param ?string $name
      * @param PropertyValueGenerator|string|array|null $defaultValue
-     * @param int $flags
+     * @param int|int[] $flags
```
Because in the tests an int array is used.

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -27,9 +27,9 @@ abstract class AbstractMemberGenerator extends AbstractGenerator
+     * @return static
      */
-    public function setFlags($flags)
+    public function setFlags($flags): self
```
Just realized that we can't add `: self` here, because child classes may be implementing `setFlags()`, and they will break, if they didn't declare `:self`

`:static` at docblock level, as well as type refinements in docblock, are sufficient

Ocramius replied:
Practical example: https://3v4l.org/f2Tu7

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
+     * @return static
      */
-    public function addFlag($flag)
+    public function addFlag(int $flag): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
+     * @return static
      */
-    public function removeFlag($flag)
+    public function removeFlag(int $flag): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
+     * @return static
      */
-    public function setAbstract($isAbstract)
+    public function setAbstract(bool $isAbstract): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
-     * @return bool
-     */
-    public function isAbstract()
+    public function isAbstract(): bool
```
Similar to above: we can't add `:bool`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
+     * @return static
      */
-    public function setInterface($isInterface)
+    public function setInterface(bool $isInterface): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
-     * @return bool
-     */
-    public function isInterface()
+    public function isInterface(): bool
```
Similar to above: we can't add `:bool`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
+     * @return static
      */
-    public function setFinal($isFinal)
+    public function setFinal(bool $isFinal): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
-     * @return bool
-     */
-    public function isFinal()
+    public function isFinal(): bool
```
Similar to above: we can't add `:bool`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
+     * @return static
      */
-    public function setStatic($isStatic)
+    public function setStatic(bool $isStatic): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
-     * @return bool
-     */
-    public function isStatic()
+    public function isStatic(): bool
```
Similar to above: we can't add `:bool`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
+     * @return static
      */
-    public function setVisibility($visibility)
+    public function setVisibility(string $visibility): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -45,98 +45,80 @@ public function setFlags($flags)
     /**
-     * @param  string $visibility
-     * @return AbstractMemberGenerator
+     * @psalm-param static::VISIBILITY_* $visibility
```
Nice touch: this is some good type improvement :+1:

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -157,9 +139,9 @@ public function setVisibility($visibility)
+     * @psalm-return static::VISIBILITY_*
      */
-    public function getVisibility()
+    public function getVisibility(): string
```
Similar to above: we can't add `:string`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -172,29 +154,25 @@ public function getVisibility()
+     * @return static
      */
-    public function setName($name)
+    public function setName(string $name): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -172,29 +154,25 @@ public function getVisibility()
-     * @return string
-     */
-    public function getName()
+    public function getName(): string
```
Similar to above: we can't add `:string`

Ocramius commented on src/Generator/AbstractMemberGenerator.php:
```diff
@@ -172,29 +154,25 @@ public function getVisibility()
+     * @return static
      */
-    public function setDocBlock($docBlock)
+    public function setDocBlock($docBlock): self
```
Similar to above: we can't add `:self`

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -4,13 +4,17 @@
 class PropertyGenerator extends AbstractMemberGenerator
 {
     public const FLAG_CONSTANT = 0x08;
+    public const FLAG_READONLY = 0x80;
```
I wonder if we should set this to `private` :thinking:

michaelpetri replied:
Totally, changed!

michaelpetri replied:
Reverted since we refer to it from tests.

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -4,13 +4,17 @@
 
+use function array_reduce;
+use function method_exists;
+use function PHPUnit\Framework\assertIsBool;
```
This symbol comes from the test suite: we can't use it in sources

michaelpetri replied:
:sweat_smile: 
Removed and replaced by condition and exception.

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -112,18 +121,21 @@ public static function fromArray(array $array)
                     $property->omitDefaultValue($value);
                     break;
+                case 'readonly':
+                    assertIsBool($value, 'Readonly must be of type bool');
```
`assert(is_bool($value), ...)`

michaelpetri replied:
Used condition with `Laminas\Code\Generator\Exception\InvalidArgumentException`, also okay?

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -136,40 +148,62 @@ public function __construct($name = null, $defaultValue = null, $flags = self::F
-     * @return PropertyGenerator
-     */
-    public function setConst($const)
+    public function setConst(bool $const): self
```
`:self` cannot be added here.

`@return static` should probably be kept at docblock level?

michaelpetri replied:
Reverted all signature changes but kept `@return static`

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -136,40 +148,62 @@ public function __construct($name = null, $defaultValue = null, $flags = self::F
+        return $this->removeFlag(self::FLAG_CONSTANT);
+    }
+
+    public function isConst(): bool
```
~~I would avoid expanding the public API here - let's get rid of this new method~~

EDIT: I see that this method is pre-existing - all good here, but we can't add a native `:bool` to it

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -136,40 +148,62 @@ public function __construct($name = null, $defaultValue = null, $flags = self::F
+        return (bool) ($this->flags & self::FLAG_CONSTANT);
+    }
+
+    public function setReadonly(bool $readonly): self
```
We should probably document `/** @return static */`

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -136,40 +148,62 @@ public function __construct($name = null, $defaultValue = null, $flags = self::F
+        return $this->removeFlag(self::FLAG_READONLY);
+    }
+
+    public function isReadonly(): bool
```
We should probably get rid of this public API - let's consider `#generate()` our main API, while reading state is probably not relevant (and not the use-case for this class)

michaelpetri replied:
We use it in tests, also this version provides such an interface for all other modifiers, imo we should keep it. Also keep in mind this is just a bool, so not much which can be misused?! But your decision!

Ocramius replied:
Remember that any API we add is API we have to maintain long-term too.

No big issue here though, let's add it 👍

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -136,40 +148,62 @@ public function __construct($name = null, $defaultValue = null, $flags = self::F
+     * {@inheritDoc}
      */
-    public function isConst()
+    public function setFlags($flags): self
```
`: self` cannot be added here

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -179,19 +213,15 @@ public function setDefaultValue(
-     * @return ?PropertyValueGenerator
-     */
-    public function getDefaultValue()
+    public function getDefaultValue(): ?PropertyValueGenerator
```
We can't add the native return type declaration here

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -179,19 +213,15 @@ public function setDefaultValue(
-     * @return string
      */
-    public function generate()
+    public function generate(): string
```
We can't add the native return type declaration here

michaelpetri replied:
Reverted and added psalm non-empty-string

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -179,19 +213,15 @@ public function setDefaultValue(
 
     /**
      * @throws Exception\RuntimeException
-     * @return string
```
One thing we can add is `/** @psalm-return non-empty-string */`

Ocramius commented:
Static analysis failing due to parsing: I think I need to set up dependabot here 🤔

michaelpetri commented:
> Static analysis failing due to parsing: I think I need to set up dependabot here thinking

I assume the problem here is that psalm runs on php80 and can not parse the test assets which contain php81 syntax?
Maybe the files just need to be excluded. Will check later when we have our half day for research.

Commit:
Added support for final const properties

Signed-off-by: Michael Petri <mpetri@lyska.io>

michaelpetri commented:
> > Static analysis failing due to parsing: I think I need to set up dependabot here thinking
> 
> I assume the problem here is that psalm runs on php80 and can not parse the test assets which contain php81 syntax? Maybe the files just need to be excluded. Will check later when we have our half day for research.

The problem is indeed he PHP version `<8.1.0`, unfortunately excluding the file via psalm.xml is not working. As discussed in slack i dropped the Class from code base and replaced it by an `eval()` to generate it at runtime.

Ocramius reviewed (approved):
LGTM :ship:

Ocramius commented:
HMM, static analysis still failed :thinking: 

```
Run vendor/bin/psalm  --output-format=github --shepherd --stats
Scanning files...
Analyzing files...

░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  60 / 185 (32%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 120 / 185 (64%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 180 / 185 (97%)
░░░░░
Error: Syntax error, unexpected T_STRING, expecting T_VARIABLE on line 9 (see https://psalm.dev/173)
🐑 results sent to shepherd.dev 🐑
Error: Process completed with exit code 2.
```

michaelpetri commented:
> HMM, static analysis still failed :thinking: 
> 
> ```
> Run vendor/bin/psalm  --output-format=github --shepherd --stats
> Scanning files...
> Analyzing files...
> 
> ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  60 / 185 (32%)
> ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 120 / 185 (64%)
> ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 180 / 185 (97%)
> ░░░░░
> Error: Syntax error, unexpected T_STRING, expecting T_VARIABLE on line 9 (see https://psalm.dev/173)
> 🐑 results sent to shepherd.dev 🐑
> Error: Process completed with exit code 2.
> ```

Okay that's strange, will try to reproduce local.

michaelpetri commented:
Seems like i removed my own commit after the soft reset / rebase issues i faced the last time :man_shrugging: 

Anyway, removed class again, added eval to generate it while runtime. Unfortunately i was not able to solve this issue:
```
ERROR: MixedAssignment - test/Generator/PropertyGeneratorTest.php:399:9 - Unable to determine the type that $generator is being assigned to (see https://psalm.dev/032)
        $generator = PropertyGenerator::fromReflection($reflectionProperty);
```
So i decided to add it to baseline, any ideas? 

I also realized that psalm is only running on PHP 8.0 in CI, should we extend matrix here as well?

Ocramius reviewed (approved):
LGTM 🚢 

Thanks @michaelpetri!
